File: 上學期期末考/[2017AF-E]解開一個100位的質數問題要算到世界末日.cpp
#include <bits/stdc++.h>
using namespace std;

#define miyuki_is_my_wife_and_sister ios::sync_with_stdio(false), cin.tie(nullptr)

void sieve(std::vector<int>& prime, const int n) {
    vector<bool> vis(n, false);
    for (int i = 2; i <= n; ++i) {
        if (!vis[i]) {
            prime.push_back(i);
        }
        for (const auto &p : prime) {
            if (i * p > n) {
                break;
            }
            vis[i * p] = true;
            if (i % p == 0) {
                break;
            }
        }
    }
}

int main() {
    miyuki_is_my_wife_and_sister;

    std::vector<int> prime;
    sieve(prime, 10000); //我估計10000就夠了

    int n{}, cnt{1};
    cin >> n;
    for (auto &p : prime) {
        if (p >= n) {
            break;
        }
        cout << "NOP(" << cnt++ << ")=" << p << '\n';
    }

    return 0;
}

File: 上學期期末考/[2017AF-E]解開一個100位的質數問題要算到世界末日.py
def main():
    def sieve(n):
        vis = [False] * n
        prime = []
        for i in range(2, n):
            if not vis[i]:
                prime.append(i)
            
            for p in prime:
                if i * p >= n:
                    break
                
                vis[i * p] = True
                if not i % p:
                    break
                    
        return prime
        
    prime = sieve(10000)
    
    n = int(input())
    c = 1
    
    for p in prime:
        if p >= n:
            break
        
        print("NOP({})={}".format(c, p))
        c += 1

if __name__ == "__main__":
    # miyuki is my wife and sister

    main()
